src/NodeApi/JSDispatcherQueue.cs

```diff
@@ -0,0 +1,221 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Hermes.Example;
+
+public delegate void JSTypedEventHandler<TSender, TResult>(TSender sender, TResult args);
+
+public sealed class DispatcherQueueShutdownStartingEventArgs : EventArgs
+{
+    private Func<JSDispatcherQueueDeferral> _getDeferral;
+
+    internal DispatcherQueueShutdownStartingEventArgs(Func<JSDispatcherQueueDeferral> getDeferral)
+        => _getDeferral = getDeferral;
+
+    public JSDispatcherQueueDeferral GetDeferral() => _getDeferral();
+}
+
+public sealed class JSDispatcherQueue
+{
+    private readonly object _queueMutex = new();
+    private List<Action?> _writerQueue = new(); // Queue to add new items
+    private List<Action?> _readerQueue = new(); // Queue to read items from
```

Comment:
Can the queues' element type be non-nullable `Action`? I don't see any reason it needs to be nullable, except where line 73 sets an index to null after invoking the action, but that seems unnecessary since the queue will be cleared afterward.

Reply:
Good point! Changed. I use similar approach in my C++ to free captured task resources as soon as possible. I guess it is not a big deal in C# code because the resources cannot be freed immediately anyway.